components/raftstore-v2/src/operation/ready/apply_trace.rs

```diff
@@ -0,0 +1,592 @@
+                flushed_index,
+            },
+        );
+    }
+}
+
+/// An alias of frequent use type that each data cf has a u64.
+pub type DataTrace = [u64; DATA_CFS_LEN];
+
+#[derive(Clone, Copy, Default)]
+struct Progress {
+    flushed: u64,
+    /// The index of last entry that has modification to the CF.
+    ///
+    /// If `flushed` == `last_modified`, then all data in the CF is persisted.
+    last_modified: u64,
+}
+
+pub fn cf_offset(cf: &str) -> usize {
+    let cf = if cf.is_empty() { CF_DEFAULT } else { cf };
+    DATA_CFS.iter().position(|c| *c == cf).expect(cf)
+}
+
+/// `ApplyTrace` is used to track the indexes of modifications and flushes.
+///
+/// It has 3 core functionalities:
+/// - recover from stopped state and figure out the correct log replay start
+///   point.
+/// - trace the admin flushed index and issue persistence once admin operation
+///   is considered finished. Note only those admin commands that needs to
+///   interact with other peers will be traced.
+/// - support query the flushed progress without actually scanning raft engine,
+///   which is useful for cleaning up stale flush records.
+#[derive(Default)]
+pub struct ApplyTrace {
+    /// The modified indexes and flushed index of each data CF.
+    data_cfs: Box<[Progress; DATA_CFS_LEN]>,
+    /// The modified indexes and flushed index of raft CF.
+    ///
+    /// raft CF is a virtual CF that only used for recording apply index of
+    /// certain admin commands (like split/merge). So there is no flush at all.
+    /// The `flushed` field is advanced when the admin command doesn't need to
+    /// be replayed after restart. A write should be triggered to persist the
+    /// record.
+    admin: Progress,
+    /// Index that is issued to be written. It may not be truely persisted.
+    persisted_applied: u64,
+    /// `true` means the raft cf record should be persisted in next ready.
+    try_persist: bool,
+}
+
+impl ApplyTrace {
+    fn recover(region_id: u64, engine: &impl RaftEngine) -> Result<(Self, RegionLocalState)> {
+        let mut trace = ApplyTrace::default();
+        // Get all the recorded apply index from data CFs.
+        for (off, cf) in DATA_CFS.iter().enumerate() {
+            // There should be at least one record.
+            let i = engine.get_flushed_index(region_id, cf)?.unwrap();
+            trace.data_cfs[off].flushed = i;
+            trace.data_cfs[off].last_modified = i;
+        }
+        let i = engine.get_flushed_index(region_id, CF_RAFT)?.unwrap();
+        // Index of raft CF means all data before that must be persisted.
+        trace.admin.flushed = i;
+        trace.admin.last_modified = i;
+        trace.persisted_applied = i;
+        let applied_region_state = engine
+            .get_region_state(region_id, trace.admin.flushed)?
+            .unwrap();
+        let data_index = trace.data_index();
+        // If index is not larger than applied_index, it means some CF doesn't have any
```

Comment:
I don't know what does "index" and "applied_index" refers to here. It should be "data_index" and "admin_flushed"?